src/awkward/contents/unmaskedarray.py

```diff
@@ -184,7 +192,7 @@ def _getitem_fields(self, where, only_fields=()):
         )
 
     def _carry(self, carry, allow_lazy):
-        return UnmaskedArray(
+        return UnmaskedArray.simplified(
             self._content._carry(carry, allow_lazy), parameters=self._parameters
         )
```

Comment:
ByteMaskedArray and UnmaskedArray have the only two `_carry` operations with a `simplified` because they propagate the `_carry` to their `_content`. The other option-types and union-type don't do that, and so they shouldn't `simplify`.

Also, there are places in the code where we assume that a carried layout has an unchanged type (like in broadcasting: `if isinstance(x, UnionArray): x[mask].project(i)` where the `x[mask]` would carry and make it not a `UnionArray` anymore).